frmts/vrt/vrtprocesseddatasetfunctions.cpp

```diff
@@ -604,6 +605,196 @@ LUTProcess(const char * /*pszFuncName*/, void * /*pUserData*/,
+/*                        BandScaleOffsetData                           */
+/************************************************************************/
+
+constexpr std::array<GDALDataType, 10> BandScaleOffsetSupportedTypes{
+    GDT_Byte,  GDT_Int8,   GDT_UInt16, GDT_Int16,   GDT_UInt32,
+    GDT_Int32, GDT_UInt64, GDT_Int64,  GDT_Float32, GDT_Float64};
+
+namespace
+{
+struct BandScaleOffsetData
+{
+    BandScaleOffsetData(int nInBands, const double *padfNoData,
+                        CSLConstList papszFunctionArgs)
+        : m_adfInNoData(padfNoData, padfNoData + nInBands),
+          m_adfOffsets(nInBands), m_adfScales(nInBands)
+    {
+        for (const auto &[pszKey, pszValue] :
+             cpl::IterateNameValue(papszFunctionArgs))
+        {
+            if (EQUALN(pszKey, "scale_", 6))
+            {
+                int i = std::atoi(pszKey + 6);
+                m_adfScales[i - 1] = CPLAtof(pszValue);
+            }
+            else if (EQUALN(pszKey, "offset_", 7))
+            {
+                int i = std::atoi(pszKey + 7);
+                m_adfOffsets[i - 1] = CPLAtof(pszValue);
+            }
+        }
+    }
+
+    template <typename T>
+    void applyScaleOffset(size_t nElts, const void *pInBuffer,
+                          void *pOutBuffer) const
+    {
+        constexpr double nan = std::numeric_limits<double>::quiet_NaN();
+
+        const T *CPL_RESTRICT paSrc = static_cast<const T *>(pInBuffer);
+        double *CPL_RESTRICT padfDst = static_cast<double *>(pOutBuffer);
+
+        for (std::size_t iElt = 0; iElt < nElts; ++iElt)
+        {
+            for (size_t iBand = 0; iBand < m_adfOffsets.size(); iBand++)
+            {
+
+                double dfSrc = static_cast<double>(*paSrc++);
+                *padfDst++ =
+                    (dfSrc == m_adfInNoData[iBand])
+                        ? nan
+                        : (dfSrc * m_adfScales[iBand] + m_adfOffsets[iBand]);
+            }
+        }
+    }
+
+    std::vector<double> m_adfInNoData;
+    std::vector<double> m_adfOffsets;
+    std::vector<double> m_adfScales;
+};
+}  // namespace
+
+static CPLErr
+BandScaleOffsetInit(const char * /*pszFuncName*/, void * /*pUserData*/,
+                    CSLConstList papszFunctionArgs, int nInBands,
+                    GDALDataType /* eInDT */, double *padfInNoData,
+                    int *pnOutBands, GDALDataType *peOutDT,
+                    double **ppadfOutNoData, const char * /* pszVRTPath */,
+                    VRTPDWorkingDataPtr *ppWorkingData)
+{
+    const bool bIsFinalStep = *pnOutBands != 0;
+    *peOutDT = GDT_Float64;
+    *ppWorkingData = nullptr;
+
+    if (!bIsFinalStep)
+    {
+        *pnOutBands = nInBands;
+    }
+
+    auto data = std::make_unique<BandScaleOffsetData>(nInBands, padfInNoData,
+                                                      papszFunctionArgs);
+
+    // Input NoData value is meaningless after scaling, so we set the output NoData value to NaN.
```

Comment:
Not sure using `NaN` is appropriate here or if I need to look elsewhere for a declared NoData value.

Reply:
BandAffineCombination has an optional
"   <Argument name='dst_nodata' type='double' "
        "description='Override output nodata value'/>"

Reply:
Should I add that here?

Reply:
> Should I add that here?

I don't have a strong opinion if we need to provide a override dstnodata. Just documenting that the algorithm standardize on NaN should be fine